src/FSharp.AWS.DynamoDB/TableContext.fs

```diff
@@ -1402,6 +1318,141 @@ type TableContext<'TRecord>
         else
             t.VerifyTableAsync()
 
+    member t.Transaction() =
+        match metricsCollector with
+        | Some metricsCollector -> Transaction(metricsCollector = metricsCollector)
+        | None -> Transaction()
+
+/// <summary>
+///    Represents a transactional set of operations to be applied atomically to a arbitrary number of DynamoDB tables.
+/// </summary>
+/// <param name="metricsCollector">Function to receive request metrics.</param>
+and Transaction(?metricsCollector: (RequestMetrics -> unit)) =
+    let transactionItems = ResizeArray<TransactWriteItem>()
+    let mutable (dynamoDbClient: IAmazonDynamoDB) = null
+
+    let setClient client =
+        if dynamoDbClient = null then
+            dynamoDbClient <- client
+
+    let reportMetrics collector (tableName: string) (operation: Operation) (consumedCapacity: ConsumedCapacity list) (itemCount: int) =
+        collector
+            { TableName = tableName
+              Operation = operation
+              ConsumedCapacity = consumedCapacity
+              ItemCount = itemCount }
+
+    let returnConsumedCapacity, maybeReport =
+        match metricsCollector with
+        | Some sink -> ReturnConsumedCapacity.INDEXES, Some(reportMetrics sink)
+        | None -> ReturnConsumedCapacity.NONE, None
+
+    /// <summary>
+    ///    Adds a Put operation to the transaction.
+    /// </summary>
+    /// <param name="tableContext">Table context to operate on.</param>
+    /// <param name="item">Item to be put.</param>
+    /// <param name="precondition">Optional precondition expression.</param>
+    member this.Put<'TRecord>
+        (
+            tableContext: TableContext<'TRecord>,
+            item: 'TRecord,
+            ?precondition: ConditionExpression<'TRecord>
+        ) : Transaction =
+        setClient tableContext.Client
+        let req = Put(TableName = tableContext.TableName, Item = tableContext.Template.ToAttributeValues item)
+        precondition
+        |> Option.iter (fun cond ->
+            let writer = AttributeWriter(req.ExpressionAttributeNames, req.ExpressionAttributeValues)
+            req.ConditionExpression <- cond.Conditional.Write writer)
+        transactionItems.Add(TransactWriteItem(Put = req))
+        this
```

Comment:
unless there is an actual immutable builder thing going on, its confusing to return values that have to be `ignore`d

Reply:
I think it’s mostly there because that’s what people generally do with C# builders. We haven’t called it a builder so happy to drop this.